### PR TITLE
Affinity propagation

### PR DESCRIPTION
**Pull request by cjekal-uscis**
Getting Affinity Propagation and DBSCAN to work.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -0,0 +1 @@
+*.pdf
```
Comment (cjekal-uscis) [inline .gitignore:1]: figured it's not worth committing the PDFs themselves. We can remove this if you want to commit and store PDFs.

File: affinity_propagation_demo.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
+# for data prep, clustering and evaluation
+from sklearn.cluster import AffinityPropagation
+from sklearn import metrics
+import numpy as np
+
+# for plotting
+import matplotlib.pyplot as plt
+from matplotlib.pyplot import cm
+
+def create_data(nobs, vert_scale, horz_scale):
```
Comment (cjekal-uscis) [inline affinity_propagation_demo.py:10]: this was copied straight from your code @joesweetsox

File: affinity_propagation_demo.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+if __name__ == '__main__':
+    # #############################################################################
+    # Generate sample data
+    X, labels_true = create_data(500,3,1)
+    # #############################################################################
+    # Compute Affinity Propagation
+    af = AffinityPropagation(max_iter=500000, convergence_iter=2, random_state=1, verbose=True).fit(X)
+    cluster_centers_indices = af.cluster_centers_indices_
+    labels = af.labels_
```
Comment (cjekal-uscis) [inline affinity_propagation_demo.py:37]: this runs Affinity Propagation and we grab the labels

File: affinity_propagation_demo.py [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+    print('Estimated number of clusters: %d' % n_clusters_)
+    print("Homogeneity: %0.3f" % metrics.homogeneity_score(labels_true, labels))
+    print("Completeness: %0.3f" % metrics.completeness_score(labels_true, labels))
+    print("V-measure: %0.3f" % metrics.v_measure_score(labels_true, labels))
+    print("Adjusted Rand Index: %0.3f"
+        % metrics.adjusted_rand_score(labels_true, labels))
+    print("Adjusted Mutual Information: %0.3f"
+        % metrics.adjusted_mutual_info_score(labels_true, labels))
+    print("Silhouette Coefficient: %0.3f"
+        % metrics.silhouette_score(X, labels, metric='sqeuclidean'))
```
Comment (cjekal-uscis) [inline affinity_propagation_demo.py:50]: These print out some evaluation metrics

File: affinity_propagation_demo.py [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+        cluster_center = X[cluster_centers_indices[k]]
+        plt.plot(X[class_members, 0], X[class_members, 1], '.', c=col)
+        plt.plot(cluster_center[0], cluster_center[1], 'o', markerfacecolor=col,
+                markeredgecolor='k', markersize=14)
+        for x in X[class_members]:
+            plt.plot([cluster_center[0], x[0]], [cluster_center[1], x[1]], c=col)
+
+    plt.title('Estimated number of clusters: %d' % n_clusters_)
+    plt.savefig("affinity-propagation.pdf")
+    plt.show()
```
Comment (cjekal-uscis) [inline affinity_propagation_demo.py:71]: this saves the plot as PDF and shows it

File: cluster_tests_4.py [hunk 6]
```diff
@@ -30,7 +30,7 @@ def create_data(nobs, vert_scale, horz_scale):
     y_data=g1_y.tolist()+g2_y.tolist()+g3_y.tolist()+g4_y.tolist()
     data=np.array([x_data,y_data])   
     
-    fig=plt.figure(dpi=600)
+    fig=plt.figure(dpi=300)
```
Comment (cjekal-uscis) [inline cluster_tests_4.py:33]: lowered the DPI b/c @ 600 it was really big and hard to see anything

File: dbscan_demo.py [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+    x_data = g1_x.tolist()+g2_x.tolist()+g3_x.tolist()+g4_x.tolist()
+    y_data = g1_y.tolist()+g2_y.tolist()+g3_y.tolist()+g4_y.tolist()
+    return (np.array([list(a) for a in zip(x_data, y_data)]), [0 for a in range(len(g1_x))] + [1 for a in range(len(g2_x))] + [2 for a in range(len(g3_x))] + [3 for a in range(len(g4_x))])
+
+def dbscan_attempt(X, labels_true, eps):
+    # Compute DBSCAN
+    db = DBSCAN(eps=eps, min_samples=5, metric='euclidean', algorithm='auto', n_jobs=-1).fit(X)
+    core_samples_mask = np.zeros_like(db.labels_, dtype=bool)
+    core_samples_mask[db.core_sample_indices_] = True
+    labels = db.labels_
```
Comment (cjekal-uscis) [inline dbscan_demo.py:33]: This is how you run DBSCAN clustering algo

File: dbscan_demo.py [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+    plt.savefig(f"dbscan_{eps:0.3f}.pdf")
+
+if __name__ == '__main__':
+    # #############################################################################
+    # Generate sample data
+    X, labels_true = create_data(500,3,1)
+    # #############################################################################
+
+    for eps in np.linspace(0.005, 0.2, 40):
+        dbscan_attempt(X, labels_true, eps)
```
Comment (cjekal-uscis) [inline dbscan_demo.py:90]: Try out DBSCAN in 0.005 value increments for the EPS hyperparameter

**Review by joesweetsox — APPROVED** [review]
